web/source/keymanweb.js

```diff
@@ -3058,15 +3076,91 @@
 /**
  * Install a keyboard script that has been downloaded from a keyboard server
  * 
- *  @param  {string}  kbdFile   keyboard filename
+ *  @param  {Object}  kbdStub   keyboard filename
  *    
  **/      
-  keymanweb.installKeyboard = function(kbdFile)
+  keymanweb.installKeyboard = function(kbdStub)
   {
     var Lscript = util._CreateElement('SCRIPT');
     Lscript.charset="UTF-8";        // KMEW-89
-    Lscript.src = keymanweb.getKeyboardPath(kbdFile);       
     Lscript.type = 'text/javascript';
+
+    var kbdFile = kbdStub['KF'];
+    var kbdLang = kbdStub['KL'];
+    var kbdName = kbdStub['KN'];
+
+    if(typeof(kbdStub.asyncLoader.callback) == "function") {
+      // Add a handler for cases where the new <script> block fails to load.
+      Lscript.addEventListener('error', function() {
+        // Clear the timeout timer.
+        window.clearTimeout(kbdStub.asyncLoader.timer);
+        kbdStub.asyncLoader.timer = null;
+
+        // We already know the load has failed... why wait?
+        kbdStub.asyncLoader.callback('Cannot find the ' + kbdName + ' keyboard for ' + kbdLang + '!');
+        console.log('Error:  cannot find the', kbdName, 'keyboard for', kbdLang, 'at', kbdFile + '!');
+
+        kbdStub.asyncLoader = null;
+      }, true);
+    }
+    
+    // The load event will activate a newly-loaded keyboard if successful and report an error if it is not.
+    Lscript.addEventListener('load', function() {
+      if(typeof(kbdStub.asyncLoader.callback) == "function") {
+        // Clear the timeout timer.
+        window.clearTimeout(kbdStub.asyncLoader.timer);
+        kbdStub.asyncLoader.timer = null;
+      }
+
+      // To determine if the load was successful, we'll need to check the keyboard array for our desired keyboard.
+      //FIXME:
+      // Test if keyboard already loaded
+      var kbd, Li;
+      for(Li=0; Li<keymanweb._Keyboards.length; Li++) {
+        if(kbdStub['KI'] == keymanweb._Keyboards[Li]['KI']) {
+          kbd = keymanweb._Keyboards[Li];
+        } 
+      }
+
+      if(kbd != undefined) {  // Is cleared upon a successful load.
+        
+        //Activate keyboard, if it's still the active stub.
+        if(kbdStub == keymanweb._ActiveStub) {
```

Comment:
This activation code ensures that we only activate the last 'set' keyboard call in the case that we have multiple pending keyboard loads that were triggered via `SetActiveKeyboard`.  We can't really refactor it as a result, as otherwise the implementer of KeymanWeb wouldn't get deterministic behavior in such a use case.

In other words, the idea is that should we ever explicitly support multiple parallel keyboard loading calls, not all of them should be explicitly activated - only the one that was last set.